### ISSUE TITLE
Tuple with element after spread not assignable to tuple with rest elements

### ISSUE THREAD

**Issue by AnyhowStep** [microsoft/TypeScript#33573]
**TypeScript Version:**  3.5.1, 3.6.3, Nightly (2019-09-24)

**Search Terms:**

Tuple, spread, assignability

**Code**

```ts
declare const src : [number, ...number[]];

const [head, ...tail] = src;

//Expected: OK
//Actual  : OK
const dst0 : [number, ...number[]] = [head, ...tail];

//Expected: OK
//Actual  : Error
//Property '0' is missing in type 'number[]' but required in type '[number, ...number[]]'.
const dst1 : [number, ...number[]] = [head, ...tail, 1337];
```

**Expected behavior:**

Assignment to `dst1` should succeed.

It should succeed because `head` is `number` and `...tail, 1337` should correspond to `...number[]`

**Actual behavior:**

Assignment to `dst1` fails.

**Playground Link:** 

[3.5.1 Playground](http://www.typescriptlang.org/play/?ts=3.5.1#code/CYUwxgNghgTiAEYD2A7AzgF3mmZ4C54BtFAVwFsAjEGAGngDomyqaiBddgbgCgfl0WIgAsQUYPSYMMUAJYR28ALzZcvHgHoNAUQAeAB3AYQwQgHkA0po0BBMBlJQI8AvEv9UmeMEwAGVyQU1HSMzEFsnMrEouKSTDLy3Hxaeob2JuZWWnYOTi6E2jAwSDDWAArFhjAYAJ7wAOS+9fCyaPDkrWiyKADmLSjwtYYNLMEczZSkWHAAjqSycMD9gzXD9YGsIVKjEez1DB6C3pgAjAE7W2GbHIoqImISodJyEPQnAMzvAOzcQA)

**Related Issues:** 

Possibly this,
https://github.com/microsoft/TypeScript/issues/29248

It mentions spread arguments, and an extra argument at the end causes TS to not realize there are the right number of arguments.
However, this repro just has tuples not being assignable to other tuples.

**Comment by AnyhowStep** [comment]
This is my current workaround,
```ts
const src : readonly [number, ...number[]] = [1,2,3,4,5] as const;

interface ReadonlyArray<T> {
    concat (
        this : readonly [T, ...T[]],
        ...items: (T | ConcatArray<T>)[]
    ) : ([T, ...T[]])
}

//Expected: OK
//Actual  : OK
const dst2 : [number, ...number[]] = src.concat(1337);

console.log(src);
console.log(dst2);
```

[3.5.1 Playground](http://www.typescriptlang.org/play/?ts=3.5.1#code/MYewdgzgLgBBBOwYC4bwKYEMAm4A2AnjANpgCuAtgEbrwA0MAdM+dbcQLocwC8JAjHQBMdAMx0ALHQCs3TBBihIUANwAoNQEswUWgDNMwdDABKWXGEIBBePEwEAPABUAfDADeamN8XhgmWAAKLx9QqAALTQVUDBx8ImInBmZGJ04OOhDQ7xTNXQoIVECnGAAfGABhPwCbO0dXAEpOLJgGlBhAxOTmNK4GtQBfDQB6YYBRAA8AB3RgXWxUAHkAaTVRqzmyTDxvJdWlaBhsaCF20koaeiYWC-YuXjhERiV-KED+UVEAdgb1NQOQHh0Iw8CAAOaBBDAX7-cAQQHA0EQ45QIS-IA)

-----

If anyone is using this for a tuple-of-arrays, you'll need,
```ts
const src : readonly [[number, number], ...[number, number][]] = [[1,2], [3,4]];

interface ReadonlyArray<T> {
    concat (
        this : readonly [T, ...T[]],
        ...items: (T | ConcatArray<T>)[]
    ) : ([T, ...T[]])
}
interface Array<T> {
    concat (
        this : readonly [T, ...T[]],
        ...items: (T | ConcatArray<T>)[]
    ) : ([T, ...T[]])
}

//Expected: OK
//Actual  : OK
const dst2 : readonly [[number, number], ...[number, number][]] = src.concat([5,6]);

console.log(src);
console.log(dst2);
```
[3.5.1 Playground](http://www.typescriptlang.org/play/?ts=3.5.1#code/MYewdgzgLgBBBOwYC4bwKYEMAm4A2AnjANrFgCuAtgEbrwA0MFNdAuowHRdlW0NO82xVqxgBeEsQCM9AEzsSAZnoAWEQG4AUJoCWYKHQBmmYOhgAlLLjCEAgvHiYCAHgAqAPhgBvTTD8xQMGBMWAAKX39IqAALHQgUNCt8ImJXTi5XYXYIyL8uDh0DSghUUNcYAB8YAGFwYKh7RxcPAEphHJgWhNDU9I5MkRbNAF9dfSMTM0anN08fSMD6mHDc-xi4hIwcZJI0mHyB7NW8rkL0YtLyqtqgkOnm9zbWDq7S3v2MrKHRzQB6X4AogAPAAO6GABmwqAA8gBpP6-WwQ8iYPB+GHwwLQGDYaCyTZJGwpHgsfjMPgKfIkviMclCETiOCIDiLEI9ACs9AAbKwWlpNFiQHh0Bw8CAAOahBDAPkC8AQIUisWS3FQWR8oA)

**Comment by Andarist** [comment]
@RyanCavanaugh this works OK today, the issue can be closed

**Comment by jakebailey** [comment]
Just looking at old issues; this was fixed in #39094.
